### ISSUE TITLE
[QST] SegFault when performing TiledCopy

### ISSUE THREAD

**Issue by phantaurus** [NVIDIA/cutlass#1735]
**What is your question?**
Hello, the following code generates SegFault during compilation but the code looks ok. Could you help me take a look at this issue? Appreciate your help!

I am basically just doing DefaultCopy here, which copies only one value each time. Therefore, it doesn't seem that it is because of a layout mismatch...

```
auto tiled_mma = TiledMMA<MMA_Atom<SM80_16x8x16_F32F16F16F32_TN>,
                            Layout<Shape<8, _1, _1>>,
                            Tile<_128, _16, _16>>{};
auto smem_copy_atom = Copy_Atom<DefaultCopy, cutlass:half_t>{};
auto smem_tiled_copy = make_tiled_copy_C(smem_copy_atom, tiled_mma);
auto smem_thr_copy = smem_tiled_copy.get_thread_slice(tidx);

// ... Creating the input tensor rO and sO
// rO has layout ((_2,_2),_1,_16):((_1,_2),_0,_4)
// sO has layout (_128,(_32,_4)):(_32,(_1,_4096))

Tensor tOrO = smem_thr_copy.retile_S(rO);
Tensor tOsO = smem_thr_copy.partition_D(sO);

// tOrO has layout ((_1,_8),_1,_8):((_0,_1),_0,_8)
// tOsO has layout ((_1,(_2,_2,_2)),_1,(_2,_4)):((_0,(_1,_256,_8)),_0,(_16,_4096))

cute::copy(smem_tiled_copy, tOrO, tOsO);

```

The cute::copy here generates Segmentation Fault during compilation. The program returns fine if this line is commented out.

**Comment by ccecka** [comment]
That's unfortunate -- I'm currently auditing and improving the `copy` utilities in CuTe to prevent this in the future.

Looks to me like your `r0` is the output from `MMA_Atom`? Then it would have `value_type` of F32, so have you tried this
```cpp
using T = typename decltype(r0)::value_type;
auto smem_copy_atom = Copy_Atom<UniversalCopy<T>, T>{};
```

**Comment by phantaurus** [comment]
Thank you so much for your reply! rO has been converted to fp16 before the copy.
```
// acc_o is cute::gemm output
// convert_to is a wrapper of NumericArrayConverter
Tensor rO = convert_to<cutlass::half_t>(acc_o);
```

**Comment by ccecka** [comment]
Ok, I would still try `UniversalCopy` instead of `DefaultCopy` and see if that fixes it.

**Comment by phantaurus** [comment]
I tried copy directly from rO to gO:
```
// gO has layout (_128,_128):(128,_1)
Tensor tOgO = thr_mma.partition_C(gO);

// tOgO has layout ((_2,_2),_1,_16):((_1,1024),_0,_8)
// rO has layout ((_2,_2),_1,_16):((_1,_2),_0,_4)
cute::copy(rO, tOgO);
```
This also generates SegFaults occasionally... However, sometimes it compiles. Commenting out the cute::copy line also seems to solve this issue.

**Comment by phantaurus** [comment]
I also tried this, but the error persists
```
using SmemCopyAtomO = Copy_Atom<UniversalCopy<cutlass::half_t>, cutlass:half_t>;
```

Just wondering if __syncthreads() or cp async wait calls would affect the compilation of cute::copy() line? Maybe I inserted these statements improperly in my code...

**Comment by thakkarV** [comment]
What's your CUDA toolkit version?

**Comment by phantaurus** [comment]
I am using Cuda 11.6.

**Comment by phantaurus** [comment]
Oddly if I add the following print line anywhere before cute::copy, the code compiles.
```
if(thread0()) {
        print(rO);
    }
```

But printing only the layout doesn't help:
```
if(thread0()) {
        print(rO.layout());
    }
```

**Comment by phantaurus** [comment]
Actually print acc_o (the FP32 version of rO) also works...

My rO (fp16) is converted from acc_o (fp32), which is the cute::gemm output:
```
Tensor acc_o = partition_fragment_C(tiled_mma, Shape<Int<kBlockM>, Int<kDimX>>{});
cute::gemm(tiled_mma, ..., acc_o);
Tensor rO = convert_to<cutlass:half_t>(acc_o);
```

The convert function is copied directly from flash attention:
```
template <typename To_type, typename Engine, typename Layout>
inline __device__ auto convert_to(Tensor<Engine, Layout> const& tensor) {
  using From_type = typename Engine::value_type;
  constexpr int numel = decltype(size(tensor))::value;
  cutlass::NumericArrayConverter<To_type, From_type, numel> convert_op;
  auto frag =
      convert_op(*reinterpret_cast<const cutlass::Array<From_type, numel>*>(
          tensor.data()));
  return make_tensor(make_rmem_ptr<To_type>(&frag), tensor.layout());
}
```

 I didn't specifically look into the convert_to before. I used this function in my previous kernels and they all run successfully.

**Comment by ccecka** [comment]
Wow, that looks super dangerous to me.
1. `reinterpret_cast<const cutlass::Array<From_type, numel>*>(tensor.data())`: assumes that the input tensor is compact and that `From_type` isn't subbyte.
2. `make_rmem_ptr<To_type>(&frag)`: is a dangling pointer to `cutlass::Array<To_type, N>`.
3. `make_tensor(ptr, tensor.layout());`: Also assumes that `layout` is compact.

Try something trivial first instead of `convert_to`:
```cpp
Tensor r0 = make_tensor_like<cutlass::half_t>(acc_o);
cute::copy(acc_o, r0);
```

**Comment by phantaurus** [comment]
Thanks! I have changed to use cute:copy directly. However, the issue persists. I still have to print(rO) in order to compile successfully.

**Comment by thakkarV** [comment]
Please update your toolkit to something newer. At least 11.8. We do not support 11.6 with CUTLASS 3

**Comment by phantaurus** [comment]
This issue is resolved after updating to 11.8. Really appreciate your help!
